### PR TITLE
Resolve userId with resident organization

### PR DESCRIPTION
**Pull request by sahandilshan**
### Proposed changes in this pull request
- If the userId is empty for the given tenant, check if the user is available in resident org and try to resolve it

This will fix the https://github.com/wso2/product-is/issues/17915

### CODE REVIEW

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/IdentityCoreServiceComponent.java [hunk 1]
```diff
@@ -359,4 +360,30 @@ public static ServiceURLBuilderFactory getServiceURLBuilderFactory() {
...
+            name = "identity.organization.management.component.resident.resolver",
+            service = OrganizationUserResidentResolverService.class,
+            cardinality = ReferenceCardinality.OPTIONAL,
+            policy = ReferencePolicy.DYNAMIC,
+            unbind = "unsetOrganizationUserResidentResolverService"
+    )
+    protected void setOrganizationUserResidentResolverService(
+            OrganizationUserResidentResolverService organizationUserResidentResolverService) {
+
+        if (log.isDebugEnabled()) {
```
Comment (sadilchamishka) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/IdentityCoreServiceComponent.java:374]: Debug log check not required. Better update the license also.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java [hunk 2]
```diff
@@ -1648,7 +1652,24 @@ public static String resolveUserIdFromUsername(int tenantId, String userStoreDom
                         if (log.isDebugEnabled()) {
                             log.debug("User id could not be resolved for username: " + username + " in user store " +
                                     "domain: " + userStoreDomain + " and tenant with id: " + tenantId + ". Probably " +
-                                    "user does not exist in the user store.");
+                                    "user does not exist in the user store. Hence trying to resolve user from the " +
+                                    "resident organization.");
+                        }
+                        Tenant tenant = IdentityCoreServiceDataHolder.getInstance().getRealmService()
+                                .getTenantManager().getTenant(tenantId);
+                        if(tenant == null) {
```
Comment (sadilchamishka) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java:1660]: if (tenant == null)

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java [hunk 3]
```diff
@@ -1648,7 +1652,24 @@ public static String resolveUserIdFromUsername(int tenantId, String userStoreDom
...
+                            return userId;
+                        }
+                        String associatedOrganizationUUID = tenant.getAssociatedOrganizationUUID();
+                        if (associatedOrganizationUUID != null) {
+                            // Trying to resolve the user from the Resident Organization if the user is not found in
+                            // the user store.
```
Comment (sadilchamishka) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java:1666]: if the user is not found in the user store of the current organization.

**Review by sadilchamishka — DISMISSED** [review]

**Comment by sahandilshan** [comment]
In order to run the integration test, we need to merge https://github.com/wso2/product-is/pull/17930 PR first.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java [hunk 4]
```diff
@@ -1648,7 +1652,24 @@ public static String resolveUserIdFromUsername(int tenantId, String userStoreDom
...
+                                    "user does not exist in the user store. Hence trying to resolve user from the " +
+                                    "resident organization.");
+                        }
+                        Tenant tenant = IdentityCoreServiceDataHolder.getInstance().getRealmService()
+                                .getTenantManager().getTenant(tenantId);
+                        if(tenant == null) {
+                            return userId;
+                        }
+                        String associatedOrganizationUUID = tenant.getAssociatedOrganizationUUID();
+                        if (associatedOrganizationUUID != null) {
```
Comment (chamathns) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java:1664]: let's use String Utils

**Review by chamathns — APPROVED** [review]

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java (hunk 4 above)
Comment (sahandilshan) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/5184/commits/01717332471b2c20636862ea3ff943be0905c224

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java (hunk 3 above)
Comment (sahandilshan) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/5184/commits/01717332471b2c20636862ea3ff943be0905c224

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java (hunk 2 above)
Comment (sahandilshan) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/5184/commits/01717332471b2c20636862ea3ff943be0905c224

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/IdentityCoreServiceComponent.java (hunk 1 above)
Comment (sahandilshan) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/IdentityCoreServiceComponent.java]: Fixed 01717332471b2c20636862ea3ff943be0905c224

**Review by sadilchamishka — APPROVED** [review]

**Comment by sahandilshan** [comment]
```
org.wso2.identity.integration.test.rest.api.server.application.management.v1.ApplicationMetadataPositiveTest.testGetAdaptiveAuthTemplates(org.wso2.identity.integration.test.rest.api.server.application.management.v1.ApplicationMetadataPositiveTest)
[ERROR]   Run 1: ApplicationMetadataPositiveTest.testGetAdaptiveAuthTemplates:229 » JSON JSONOb...
[ERROR]   Run 2: ApplicationMetadataPositiveTest.testGetAdaptiveAuthTemplates:229 » JSON JSONOb...
[INFO] 
```

Only one test case is failing in the PR builder, and it's a known test failure. Hence merging this PR

**Pull request merged**